pytorch_lightning/trainer/trainer.py

```diff
@@ -1095,16 +1101,39 @@ def tune(
                 If the model has a predefined val_dataloaders method this will be skipped
 
             datamodule: A instance of :class:`LightningDataModule`.
+
+            scale_batch_size_kwargs: Arguments for :func:`~pytorch_lightning.tuner.lr_finder.lr_find`
+
+            lr_find_kwargs: Arguments for :func:`~pytorch_lightning.tuner.batch_size_scaling.scale_batch_size`
         """
         Trainer._log_api_event("tune")
         self.state = TrainerState.TUNING
         self.tuning = True
 
-        self.tuner.tune(model, train_dataloader, val_dataloaders, datamodule)
+        # if a datamodule comes in as the second arg, then fix it for the user
+        if isinstance(train_dataloader, LightningDataModule):
+            datamodule = train_dataloader
+            train_dataloader = None
+        # If you supply a datamodule you can't supply train_dataloader or val_dataloaders
+        if (train_dataloader is not None or val_dataloaders is not None) and datamodule is not None:
+            raise MisconfigurationException(
+                'You cannot pass `train_dataloader` or `val_dataloaders` to `trainer.tune(datamodule=...)`'
+            )
+
+        # links data to the trainer
+        self.data_connector.attach_data(
+            model, train_dataloader=train_dataloader, val_dataloaders=val_dataloaders, datamodule=datamodule
+        )
```

Comment:
This seems redundant. Refactor into function?

Reply:
Why do you say it's redundant?

I don't think it's worth moving into a function as it is only called here. `trainer.fit` is also similar but the error message is different.